pandas/core/indexes/multi.py

```diff
@@ -1412,6 +1420,77 @@ def from_product(cls, iterables, sortorder=None, names=None):
         labels = cartesian_product(labels)
         return MultiIndex(levels, labels, sortorder=sortorder, names=names)
 
+    @classmethod
+    def from_frame(cls, df, squeeze=True, names=None):
+        """
+        Make a MultiIndex from a DataFrame.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            DataFrame to be converted to MultiIndex.
+        squeeze : bool, default True
+            If df is a single column, squeeze MultiIndex to be a regular Index.
+        names : list / sequence / callable, optonal
+            If no names provided, use column names, or tuple of column names if
+            the columns is a MultiIndex. If sequence, overwrite names with the
+            given sequence. If callable, pass each column name or tuples of
```

Comment:
not really sure of the difference of these, can you show what the rationale for all of these options?

Reply:
The callable option was mostly for cases where the frame used to construct the mi is itself multiindexed on the columns. Example below:

        >>> df = pd.DataFrame([['a', 'b'], ['c', 'd'], ['e', 'f']],
        ...                   columns=pd.MultiIndex.from_tuples([('L1', 'x'), 
        ...                                                      ('L2', 'y')]))
        >>> df
          L1 L2
           x  y
        0  a  b
        1  c  d
        2  e  f
        >>> pd.MultiIndex.from_frame(df, names=lambda x: '_'.join(x))
        MultiIndex(levels=[['a', 'c', 'e'], ['b', 'd', 'f']],
                   labels=[[0, 1, 2], [0, 1, 2]],
                   names=['L1_x', 'L2_y'])

Reply:
This would be an uncommon occurrence. Would it make more sense to just not provide the callable option?

Reply:
let's remove the callable option for now